File: framework/languages/java/zlink-framework-core/src/main/java/systems/zlink/framework/runtime/streams/ZLinkStreamInboundFrame.java
package systems.zlink.framework.runtime.streams;

import java.util.Objects;
import java.util.concurrent.atomic.AtomicBoolean;
import systems.zlink.contracts.messaging.Message;

/** Owns the two messages produced by one complete STREAM frame. */
final class ZLinkStreamInboundFrame implements AutoCloseable {
    private Message header;
    private Message payload;
    private final Runnable terminalRelease;
    private final AtomicBoolean closed = new AtomicBoolean();

    ZLinkStreamInboundFrame(Message header, Message payload) {
        this(header, payload, () -> { });
    }

    ZLinkStreamInboundFrame(
        Message header,
        Message payload,
        Runnable terminalRelease) {
        this.header = Objects.requireNonNull(header, "header");
        this.payload = Objects.requireNonNull(payload, "payload");
        this.terminalRelease = Objects.requireNonNull(
            terminalRelease, "terminalRelease");
    }

    Message header() {
        return header;
    }

    Message payload() {
        return payload;
    }

    @Override
    public void close() {
        if (!closed.compareAndSet(false, true)) {
            return;
        }
        Message currentHeader = header;
        Message currentPayload = payload;
        header = null;
        payload = null;
        try {
            if (currentHeader != null) {
                currentHeader.close();
            }
            if (currentPayload != null) {
                currentPayload.close();
            }
        } finally {
            terminalRelease.run();
        }
    }
}

File: framework/languages/dotnet/src/Zlink.Framework/Runtime/Streams/ZLinkStreamInboundFrame.cs
using Zlink.Framework.Runtime.Execution;

namespace Zlink.Framework.Runtime.Streams;

internal sealed class ZLinkStreamInboundFrame(
    Message header,
    Message payload) : IDisposable
{
    internal Message? Header { get; private set; } = header;
    internal Message? Payload { get; private set; } = payload;

    internal ZLinkApplicationJobQueueLease? ApplicationJobAdmission { get; set; }

    internal void Detach()
    {
        Header = null;
        Payload = null;
        ApplicationJobAdmission = null;
    }

    public void Dispose()
    {
        Header?.Dispose();
        Payload?.Dispose();
        Header = null;
        Payload = null;
        ApplicationJobAdmission?.Dispose();
        ApplicationJobAdmission = null;
    }
}
